clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3007,19 +3007,37 @@ class SYCLKernelNameTypeVisitor
         S.getASTContext().getLangOpts().SYCLUnnamedLambda;
     const DeclContext *DeclCtx = Tag->getDeclContext();
     if (DeclCtx && !UnnamedLambdaEnabled) {
-      auto *NameSpace = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
-      if (NameSpace && NameSpace->isStdNamespace()) {
-        S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << KernelNameType;
-        S.Diag(KernelInvocationFuncLoc, diag::note_invalid_type_in_sycl_kernel)
-            << /* kernel name cannot be a type in the std namespace */ 2
-            << QualType(Tag->getTypeForDecl(), 0);
-        IsInvalid = true;
-        return;
+      // Loop through Declaration Contexts to detect kernel names nested inside
+      // "std" and "anonymous" namespaces.
+      while (!DeclCtx->isTranslationUnit()) {
+        auto *NSDecl = dyn_cast<NamespaceDecl>(DeclCtx);
+        if (NSDecl && NSDecl->isStdNamespace()) {
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::err_sycl_kernel_incorrectly_named)
+              << KernelNameType;
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::note_invalid_type_in_sycl_kernel)
+              << /* kernel name cannot be a type in the std namespace */ 2
+              << QualType(Tag->getTypeForDecl(), 0);
+          IsInvalid = true;
+          return;
+        }
+        if (NSDecl && NSDecl->isAnonymousNamespace()) {
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::err_sycl_kernel_incorrectly_named)
+              << KernelNameType;
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::note_invalid_type_in_sycl_kernel)
+              << /* kernel name is not globally-visible */ 0
+              << QualType(Tag->getTypeForDecl(), 0);
+          IsInvalid = true;
+          return;
+        }
+        DeclCtx = DeclCtx->getParent();
       }
       if (!DeclCtx->isTranslationUnit() && !isa<NamespaceDecl>(DeclCtx)) {
-        const bool KernelNameIsMissing = Tag->getName().empty();
-        if (KernelNameIsMissing) {
+        const bool UnnamedTypeUsed = Tag->getName().empty();
```

Comment:
FWIW, community code typically does not use top-level `const` on value types that are locals or parameters.